src/pudl/output/ferc1.py

```diff
@@ -1731,68 +1842,135 @@ def exploded_calcs_to_digraph(
         forest = nx.from_pandas_edgelist(edgelist, create_using=nx.DiGraph)
         return forest
 
-    def set_forest_attributes(
-        self,
-        forest: nx.DiGraph,
-        exploded_calcs: pd.DataFrame,
-        exploded_meta: pd.DataFrame,
-        tags: pd.DataFrame,
-    ) -> nx.DiGraph:
-        """Set the attributes of a forest."""
+    @cached_property
+    def node_attrs(self: Self) -> dict[NodeId, dict[str, dict[str, str]]]:
+        """Construct a dictionary of node attributes for application to the forest.
+
+        Note attributes consist of the manually assigned tags.
+        """
         # Reshape the tags to turn them into a dictionary of values per-node. This
         # will make it easier to add arbitrary sets of tags later on.
-        tags_dict = tags.set_index(["table_name", "xbrl_factoid"]).to_dict(
-            orient="index"
-        )
-        tags_dict_df = pd.DataFrame(
-            index=pd.MultiIndex.from_tuples(
-                tags_dict.keys(), names=["table_name", "xbrl_factoid"]
-            ),
-            data={"tags": list(tags_dict.values())},
-        ).reset_index()
-
-        multi_valued_weights = (
-            exploded_calcs.groupby(self.calc_cols, dropna=False)["weight"]
-            .transform("nunique")
-            .gt(1)
+        tags_dict = (
+            self.tags.convert_dtypes().set_index(self.calc_cols).to_dict(orient="index")
         )
-
-        calcs_to_drop = multi_valued_weights & (exploded_calcs.weight == 1)
-
-        logger.info(
-            f"Found {len(calcs_to_drop)}/{len(exploded_calcs)} calculations "
-            "where weight was both 1 and -1. Dropping the 1-weighted "
-            "calculations."
+        node_attrs = (
+            pd.DataFrame(
+                index=pd.MultiIndex.from_tuples(tags_dict.keys(), names=self.calc_cols),
+                data={"tags": list(tags_dict.values())},
+            )
+            .reset_index()
+            # Type conversion is necessary to get pd.NA in the index:
+            .astype({col: pd.StringDtype() for col in self.calc_cols})
+            # We need a dictionary for *all* nodes, not just those with tags.
+            .merge(
+                self.exploded_meta.loc[:, self.calc_cols],
+                how="right",
+                on=self.calc_cols,
+                validate="one_to_many",
+            )
+            # For nodes with no tags, we assign an empty dictionary:
+            .assign(tags=lambda x: np.where(x["tags"].isna(), {}, x["tags"]))
+            .set_index(self.calc_cols)
+            .to_dict(orient="index")
         )
+        return node_attrs
 
-        attr_cols = ["weight"]
-        deduped_calcs = exploded_calcs[~calcs_to_drop][
-            self.calc_cols + attr_cols
-        ].drop_duplicates()
+    @cached_property
+    def edge_attrs(self: Self) -> dict[Any, Any]:
```

Comment:
Love this split between edges and nodes - way better than the old world!

Reply:
Yeah I don't know what I was thinking before.